MS Teams access plugin

gzigzigzeo opened the pull request:
Needs to be merged after #589 

Closes gravitational/teleport#11818

Joerger commented on lib/plugindata/cas.go:
```diff
@@ -0,0 +1,129 @@
+// Method returns modified plugin data on success, nil when data was not saved, error otherwise.
+//
+// container argument must contain the reference to an empty struct implementing PluginData interface.
+func (c *CAS) Modify(ctx context.Context, resource string, cb Cb) (interface{}, error) {
```
Can we split this into `Modify` and `Add`, where `Modify` returns a `trace.NotFound` if `oldData == nil` and `Add` returns a `trace.AlreadyExists` if `oldData != nil`? Then callers can check relevant errors instead of checking `pluginData == nil`.

I think this would be more straightforward and easy to use, which is important given this is a new library.

gzigzigzeo replied:
Compare-and-swap supposed to be a single atomic instruction. I possibly get your idea wrong, but splitting this into two methods would break the atomicity. Just in case, this method is not new. It is used in every plugin like this, I just tried to DRY it into a separate module.

Joerger replied:
I agree that we should use compare-and-swap, but we can still move the functionality from the callback to the function by splitting it into two functions. Here's what I mean:

```
func (c *CAS) Modify(ctx context.Context, resource string, modify func(interface{}) interface{}) (interface{}, error) {
	backoff := backoff.NewDecorr(c.backoffBase, c.backoffMax, clockwork.NewRealClock())
	for {
		oldData, err := c.getPluginData(ctx, resource)
		if err != nil {
			return nil, trace.Wrap(err)
		}

		newData := modify(oldData)

		err = trace.Wrap(c.updatePluginData(ctx, resource, newData, oldData))
		if err == nil {
			return newData, nil
		}
		if !trace.IsCompareFailed(err) {
			return nil, trace.Wrap(err)
		}
		if err := backoff.Do(ctx); err != nil {
			return nil, trace.Wrap(err)
		}
	}
}

func (c *CAS) Create(ctx context.Context, resource string, pd interface{}) (interface{}, error) {
	// We don't need to do backoff, because if compare fails
	// that means the plugin data already exists. Abort creation.
	err := trace.Wrap(c.updatePluginData(ctx, resource, pd, nil))
	if err != nil {
		return nil, trace.Wrap(err)
	}
	return pd, nil
}
```

Now we can simplify our calls to Modify like this:

```
// postMessages posts initial Teams messages
func (a *App) postMessages(ctx context.Context, recipients []string, id string, data pd.AccessRequestData) error { 
	...
	// Let's update sent messages data
	_, err = a.pd.Modify(ctx, id, func(existing interface{}) (interface{}) {
		pluginData, ok := existing.(PluginData)
		if !ok {
			// maybe return error?
			return nil
		}
		pluginData.TeamsData = teamsData
		return pluginData
	})
	return trace.Wrap(err)
}
```

And we can simplify Creation like this:

```
func (a *App) onPendingRequest(ctx context.Context, req types.AccessRequest) error {
	...
	pd, err := a.pd.Modify(ctx, id, PluginData{AccessRequestData: data})
	if err != nil {
		return trace.Wrap(err)
	}
```

gzigzigzeo replied:
Ah, now I see. I've already tried to achieve the same goal with `mode` argument of the method. There are actually three situations: "Create", "Update" and "Upsert". 

Let me finish my refactoring, I get back to you and we discuss an updated version.

gzigzigzeo replied:
I guess, I can close this conversation

r0mant commented on access/ms-teams/Makefile:
```diff
@@ -0,0 +1,77 @@
+VERSION=9.3.9
+GO_VERSION=1.17.11
```
Should we use the latest release now that Teleport 10 is out? It also uses Go 1.18.

gzigzigzeo replied:
With pleasure :-)

r0mant commented on access/ms-teams/app.go:
```diff
@@ -0,0 +1,517 @@
+	// pluginName used as Teleport plugin identifier
+	pluginName = "ms-teams"
+	// minServerVersion is the minimal teleport version the plugin supports.
+	minServerVersion = "6.1.0-beta.1"
```
Where did this version come from? Should probably set it to something that is supported. How about `8.0.0`?

gzigzigzeo replied:
Frankly, I do not know. Looks like that minimal version was copy pasted all over the plugins from the beginning of times. I've replaced it here, but I guess we need to create an issue to replace it everywhere else.

r0mant commented on access/ms-teams/app.go:
```diff
@@ -0,0 +1,517 @@
+	}
+
+	a.Process = lib.NewProcess(ctx)
+	// a.mainJob = lib.NewServiceJob(a.run)
```
Debug leftover?

Joerger commented on access/ms-teams/app.go:
```diff
@@ -0,0 +1,517 @@
+	log := logger.Get(ctx)
+
+	teamsApp, err := a.bot.GetTeamsApp(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if teamsApp == nil {
+		return trace.NotFound(
+			"MS Teams app not found in org app store. Please, refer to the documentation!",
+		)
+	}
```
Check not found error
```suggestion
	if err != nil {
		return trace.Wrap(err)
	}
```

r0mant commented on access/ms-teams/app.go:
```diff
@@ -0,0 +1,517 @@
+		return trace.Wrap(err)
+	}
+
+	var webProxyAddr string
+	if pong.ServerFeatures.AdvancedAccessWorkflows {
+		webProxyAddr = pong.ProxyPublicAddr
+	}
```
This logic is not super clear - why are we setting webProxyAddr if AdvancedAccessWorkflows are enabled? What if they're not enabled? The webProxyAddr will be empty and passed to NewBot below - what will happen then?

gzigzigzeo replied:
`webProxyAddr` is used to show a fancy "Open" button, which launches Teleport Web UI and opens the access request in question. As far as I understand, that feature exists in Enterprise version only - that's why we check `AdvancedAccessWorkflows`. Yes, empty value indicates to app that it should not render those links. It works like this for all other plugins, should we change it?

r0mant commented on access/ms-teams/app.go:
```diff
@@ -0,0 +1,517 @@
+
+	a.mainJob.SetReady(ok)
+
+	<-a.watcherJob.Done()
```
Should this select on the ctx.Done() also?

gzigzigzeo replied:
As far as I remember, it is triggered when the parent context is done. But, just in case, I've added it.

r0mant commented on access/ms-teams/app.go:
```diff
@@ -0,0 +1,517 @@
+	// This can happen if this set contains the channel `C` and the email for channel `C`.
+	recipientSet := stringset.New()
+
+	validEmaislSuggReviewers := []string{}
```
```suggestion
	validEmailsSuggReviewers := []string{}
```

Joerger commented on access/ms-teams/app.go:
```diff
@@ -0,0 +1,517 @@
+		RequestReason: req.GetRequestReason(),
+	}
+
+	pd, err := a.pd.Modify(ctx, id, func(existing interface{}) (interface{}, bool) {
+		if existing != nil {
+			return PluginData{}, false
+		}
+		return PluginData{AccessRequestData: data}, true
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// If we had not sent any messages before, let's create them
+	if pd != nil {
```
Can we instead use `GetPluginData` to check if it exists already, instead of creating and updating the plugin data in two steps? I imagine this could lead to an out of sync state if the initial plugin data is created, but `postMessages` fails to update the app message or fails to update the plugin data.

gzigzigzeo replied:
This would create a race condition if another instance of the plugin writes its data between get and modify calls.

Joerger replied:
We should continue to use `CompareAndSwap` to avoid such race conditions, unless that's not the kind of race condition you're referring to?

The problem I'm pointing out is that when we create the initial `PluginData{AccessRequestData: data}` object, it is essentially useless until we add the `TeamsData` in the following `Modify` call. So if `a.bot.PostMessages` fails, this plugin data will remain in this useless state, and as I understand any future calls to update it would log `"Failed to post reply: plugin data is missing"`. If we somehow had another call to `onPendingRequest`, this would also fail after seeing the plugin data already exists, even though the messages were never actually posted.

Instead, after we first `GetPluginData` just to see if it already exists. If it does, we can do `a.bot.PostMessages`. If that succeeds, we can finally do `a.pd.Modify` (or `a.pd.Create` from my other comment`) to create the full plugin data `PluginData{AccessRequestData: data, TeamsData: teamsData}`. We can always be sure that if the plugin data exists, the messages have been posted. 

This still has the possible de-synced state of the messages being posted but the plugin data failing to be created, but I'm not sure there's a simple way to circumvent this.

gzigzigzeo replied:
We can not get TeamsData unless we have messages actually sent. PluginData without TeamsData marks that an instance of the plugin has acquired a lock on this AccessRequest and is responsible for sending the initial messages. Yes, the situation when postMessages fails is possible. However, if we implement it in a way you explained, we could fall in the situation when two instances of the plugin determine that there is no plugin data and send duplicate messages.

Joerger replied:
Ah ok that makes sense, thanks for explaining

Joerger commented on access/ms-teams/app.go:
```diff
@@ -0,0 +1,517 @@
+
+// postMessages posts initial Teams messages
+func (a *App) postMessages(ctx context.Context, recipients []string, id string, data pd.AccessRequestData) error {
+	teamsData, err := a.bot.PostMessages(ctx, recipients, id, data)
+	if len(teamsData) == 0 && err != nil {
+		return trace.Wrap(err)
+	}
+	for _, data := range teamsData {
+		logger.Get(ctx).WithFields(logger.Fields{
+			"id":        data.ID,
+			"timestamp": data.Timestamp,
+			"recipient": data.RecipientID,
+		}).Info("Successfully posted to MS Teams")
+	}
+	if err != nil {
+		logger.Get(ctx).WithError(err).Error("Failed to post one or more messages to MS Teams")
+	}
```
nit: check the error in one go right after the call
```suggestion
	teamsData, err := a.bot.PostMessages(ctx, recipients, id, data)
	if err != nil {
		len(teamsData) == 0 {
			return trace.Wrap(err)
		}
		logger.Get(ctx).WithError(err).Error("Failed to post one or more messages to MS Teams")
	}
	for _, data := range teamsData {
		logger.Get(ctx).WithFields(logger.Fields{
			"id":        data.ID,
			"timestamp": data.Timestamp,
			"recipient": data.RecipientID,
		}).Info("Successfully posted to MS Teams")
	}
```

r0mant commented on access/ms-teams/config.go:
```diff
@@ -0,0 +1,67 @@
+	}
+
+	if c.MSAPI.Region == "" {
+		c.MSAPI.Region = "uk"
```
Nit: Should probably default to US (unless I'm missing what "uk" means).

gzigzigzeo replied:
That's not clear here. MS does not have proper documentation on the region codes and bot framework REST API usage. I am not even sure that this region setting affects the real region. I've changed it to `emea` as it seems the default region which is used by the .NET MS SDKs themselves.

Joerger reviewed (commented):
Mostly error handling and library usability comments, but other than that LGTM.

r0mant commented on access/ms-teams/main.go:
```diff
@@ -0,0 +1,105 @@
+	logger.Init()
+	app := kingpin.New(appName, "Teleport MS Teams plugin")
+
+	app.Command("version", "Prints teleport-slack version and exits")
```
teleport-slack -> teleport-ms-teams?

r0mant commented on access/ms-teams/main.go:
```diff
@@ -0,0 +1,105 @@
+	validateCmd := app.Command("validate", "Validate bot installation")
+	validateConfigPath := validateCmd.Flag("config", "TOML config file path").
+		Short('c').
+		Default("/etc/teleport-slack.toml").
```
teleport-ms-teams.toml?

r0mant commented on access/ms-teams/main.go:
```diff
@@ -0,0 +1,105 @@
+	startCmd := app.Command("start", "Starts Teleport MS Teams plugin")
+	startConfigPath := startCmd.Flag("config", "TOML config file path").
+		Short('c').
+		Default("/etc/teleport-slack.toml").
```
teleport-ms-teams.toml

r0mant commented on access/ms-teams/msapi/bot_framework_client.go:
```diff
@@ -0,0 +1,135 @@
+const (
+	botFrameworkAuthScope     = "https://api.botframework.com/.default"
+	botFrameworkBaseURL       = "https://smba.trafficmanager.net"
+	botFrameworkDefaultRegion = "uk"
```
Set it to "us" by default?

gzigzigzeo replied:
Unfortinately, there's no "us", and I have failed to find the region codes list. Moreover, according to my findings in the github issues, etc. bot framework is not working with all regions. So, I set this to `emea`, which seems to be the ms default region.

r0mant commented on access/ms-teams/msapi/bot_framework_client.go:
```diff
@@ -0,0 +1,135 @@
+		{
+			"type": "message",
+			"from": {
+				"id": "29:` + botID + `",
+				"name": "TeleBot"
```
What's "29"? And where does this "name" show up?

gzigzigzeo replied:
29 is an internal API designator used to determine a resource type (bot in our case). Name does not show up anywhere, it is used as a placeholder here. The problem is the following. Normally, MS bot is supposed to receive events from the users. When a user (or an admin) installs a bot for a user, the "StartConversation" event is generated. This event contains "29:<id>" along with some other information required to continue the conversation from bot's side. Bot is supposed to process this event and save all the information in the persistent storage for later usage.That's the normal way of how MS bots work. In our case there's no answering part, so I have to improvise. This approach is confirmed by an article from one of MS Teams engineers, where he proposes to do the same in case an original event got lost or a bot is totally passive and can't process any evetns.

r0mant commented on access/ms-teams/msapi/graph_client.go:
```diff
@@ -0,0 +1,265 @@
+	}
+
+	if len(users) == 0 {
+		return nil, nil
```
Should this return trace.NotFound?

r0mant commented on access/ms-teams/msapi/graph_client.go:
```diff
@@ -0,0 +1,265 @@
+	}
+
+	if len(users) == 0 {
+		return nil, nil
```
Should this return trace.NotFound?

r0mant commented on access/ms-teams/msapi/token.go:
```diff
@@ -0,0 +1,129 @@
+// extended validity window for this grant type. Hence, there is no way of preemprive refresh
+// the current token. We have to lock all requests and wait until the new token arrives.
+func (c *tokenWithTTL) Bearer(ctx context.Context, config Config) (string, error) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
```
I would try to avoid doing any I/O (incl. network calls as getToken below does) under lock. Any way to refactor this do we're not making an API request while holding a lock?

gzigzigzeo replied:
MS Graph API issues no `refresh_token` for `client_credentials` grant type. There also is no extended validity window for this grant type. Hence, there is no way of preemptive refresh the current token. We have to lock all requests and wait until the new token arrives.

I guess, that's my mistake, there's not much harm in doing multiple concurrent token refresh requests. Refactored.

r0mant commented on lib/plugindata/cas.go:
```diff
@@ -0,0 +1,129 @@
+type DecodeFn = func(map[string]string) interface{}
+
+// Cb represents a type for Plugin Data modification fn
+type Cb = func(interface{}) (interface{}, bool)
```
Can we rename this to something more descriptive? Does this stand for "callback"?

marcoandredinis commented on access/ms-teams/_tpl/manifest.json:
```diff
@@ -0,0 +1,45 @@
+{
+    "$schema": "https://developer.microsoft.com/en-us/json-schemas/teams/v1.11/MicrosoftTeams.schema.json",
+    "manifestVersion": "1.11",
```
`1.14` is the latest version
Is there a reason we are using 1.11?

gzigzigzeo replied:
That was auto generated by the ms tool. Are you sure by 100% that it would work?

marcoandredinis replied:
https://docs.microsoft.com/en-us/microsoftteams/platform/resources/schema/manifest-schema#manifestversion
Feel free to leave as 1.11 or update to 1.12

marcoandredinis commented on access/ms-teams/_tpl/manifest.json:
```diff
@@ -0,0 +1,45 @@
+        "color": "color.png",
+        "outline": "outline.png"
+    },
+    "name": {
+        "short": "TeleBot",
+        "full": "TeleBot"
+    },
+    "description": {
+        "short": "Teleport bot",
+        "full": "Teleport bot sends AccessRequests to team users"
+    },
```
@r0mant We might want to double check these descriptions with Product

marcoandredinis commented on access/ms-teams/_tpl/teleport-ms-teams.toml:
```diff
@@ -0,0 +1,14 @@
+
+[teleport]
+addr = "localhost:3025"
+identity = "identity"
```
Can we add some comments explaining how to obtain the `identity` file?
Similar to what we have in Slack

marcoandredinis commented on go.mod:
```diff
@@ -35,6 +35,11 @@ require (
 	k8s.io/apimachinery v0.24.1
 )
 
+require (
```
This should go away when we rebase with the current `master` which brings go 1.18

marcoandredinis commented:
/gcbrun

marcoandredinis commented on lib/plugindata/access_request.go:
```diff
@@ -0,0 +1,61 @@
+	Roles            []string
+	RequestReason    string
+	ReviewsCount     int
+	ResolutionTag    ResolutionTag
+	ResolutionReason string
```
:nail_care: 
The other plugins have those two fields in a single struct - `Resolution`
https://github.com/gravitational/teleport-plugins/blob/7ea82ce5d94119aeb5f23a40a8fe88edbed3d326/access/slack/plugindata.go#L14-L17
It would be good if we did that here as well, so that latter on, we can remove each individual implementation to this `/lib/plugindata`

gzigzigzeo replied:
My guess was that the separate Resolution struct her is redundant since it is almost never used separately. `/lib/plugindata` raises lots of questions, so, let me give one more pass to it and then we decide.

marcoandredinis commented on access/ms-teams/configure.go:
```diff
@@ -0,0 +1,165 @@
+
+const (
+	// TODO: Replace with live URL
+	guideURL = "http://example.com"
```
:nail_care: 
We can use `https://goteleport.com/docs/enterprise/workflow/` for now, but let's keep the TODO

gzigzigzeo commented on lib/plugindata/cas.go:
```diff
@@ -0,0 +1,129 @@
+type Cb = func(interface{}) (interface{}, bool)
+
+// CAS represents modifier struct
+type CAS struct {
```
CAS stands for compare-and-swap. Renamed it to the CompareAndSwap, as sync in go/sync package.

gzigzigzeo commented on lib/plugindata/cas.go:
```diff
@@ -0,0 +1,129 @@
+type DecodeFn = func(map[string]string) interface{}
+
+// Cb represents a type for Plugin Data modification fn
+type Cb = func(interface{}) (interface{}, bool)
```
Renamed to CallbackFn

Joerger replied:
Thanks, `ModifyFn` would also be a fitting name if you'd like.

marcoandredinis commented on lib/plugindata/cas.go:
```diff
@@ -0,0 +1,129 @@
+	UpdatePluginData(context.Context, types.PluginDataUpdateParams) error
+}
+
+// EncodeFn represents a function type of plugin data serializer
+type EncodeFn = func(interface{}) map[string]string
+
+// DecodeFn represents a function type of plugin data deserializer
+type DecodeFn = func(map[string]string) interface{}
```
(`master` is already using 1.18)

Joerger replied:
Sorry, I had meant to delete this comment once I realized this during review. Generics would be a great upgrade though, looking forward to seeing that.

marcoandredinis commented on access/ms-teams/msapi/bot_framework_client.go:
```diff
@@ -0,0 +1,135 @@
+
+// NewBotFrameworkClient creates MS Graph API client
+func NewBotFrameworkClient(config Config) *BotFrameworkClient {
+	region := config.Region
+	if region == "" {
+		region = botFrameworkDefaultRegion
+	}
+
+	baseURL := config.url.botFrameworkBaseURL
+	if baseURL == "" {
+		baseURL = botFrameworkBaseURL
+	}
+
+	baseURL = baseURL + "/" + region + "/" + botFrameworkVersion
```
You can use the `CheckAndSetDefaults` pattern

gzigzigzeo replied:
I guess, `CheckAndSetDefaults` here is applicable to the region only. Will move it to the `msapi.Config#CheckAndSetDefaults`.

marcoandredinis commented on access/ms-teams/msapi/bot_framework_client.go:
```diff
@@ -0,0 +1,135 @@
+}
+
+// GetChatMembers returns members of a chat
+func (c *BotFrameworkClient) GetChatMembers(ctx context.Context, chatID string) ([]ChatMember, error) {
```
Is this method being used?

gzigzigzeo replied:
Nope, it's the debug artifact, thanks.

marcoandredinis commented on access/ms-teams/msapi/bot_framework_client.go:
```diff
@@ -0,0 +1,135 @@
+
+// PostAdaptiveCardActivity sends an activity to the chat, content is AdaptiveCard
+func (c *BotFrameworkClient) PostAdaptiveCardActivity(ctx context.Context, botID, chatID, card, updateID string) (string, error) {
+	body := `
```
Can we convert this into a struct and call jsonMarshal?
It's an extra call, but I would rather let go build the json automatically

marcoandredinis commented on access/ms-teams/msapi/bot_framework_client.go:
```diff
@@ -0,0 +1,135 @@
+
+	request := request{
+		Method:      meth,
+		Path:        "conversations/" + chatID + "/activities/" + updateID,
```
Is this path valid if the `updateID` is `""` ?

I think we should set the `Path` the same way we are doing for the Method and SuccessCode

gzigzigzeo replied:
Yes, it is valid.

marcoandredinis commented on access/ms-teams/msg_slice_test.go:
```diff
@@ -0,0 +1,25 @@
+package main
```
This file ends with `_test.go` but doesn't actually contain any tests

gzigzigzeo replied:
Yes, but it contains the code essential for tests to work. `_test.go` indicates that this file belongs to the test domain.

marcoandredinis commented on access/ms-teams/Makefile:
```diff
@@ -0,0 +1,77 @@
+VERSION=10.0.2
+GO_VERSION=1.18.4
```
Don't forget to update `go.mod` as well, or just rebase from `master`

gzigzigzeo replied:
Done

marcoandredinis commented:
We need to update the `Makefile` at root as well

marcoandredinis commented on access/ms-teams/configure.go:
```diff
@@ -0,0 +1,164 @@
+	lib.PrintVersion(appName, Version, Gitref)
+	fmt.Println()
+
+	_, err := os.Stat(targetDir)
+	if !os.IsNotExist(err) {
+		return trace.Wrap(err)
+	}
```
Re-using an already existing directory didn't throw an error
It only printed the version and an empty line, nothing else

gzigzigzeo replied:
Fixed

marcoandredinis commented on access/ms-teams/README.md:
```diff
@@ -0,0 +1,176 @@
+
+8. There will be `app.zip` in the target directory. This archive contains MS Teams app. If you want to change the application name, URLs and other settings, please update `manifest.json` and recreate `app.zip` from `color.png`, `outline.png` and `manifest.json`.
+
+9. Upload the app to MS Teams organization store. Open "Apps", then "Manage your apps", click "Upload an app", choose "Upload an app to your org's app catalog", choose `app.zip` 
```
I think it's worth mentioning that Microsoft Teams Free does not support apps at all.

I created an app and was trying to install it to my Teams org.
There's not even an `Apps` menu when using the Free version

We can add a note at the beginning.

gzigzigzeo replied:
Was not aware of that. I have free account and it seems working for me.

gzigzigzeo replied:
I guess, you need kind of Azure subscription.

marcoandredinis replied:
Possibly :thinking: 
I created an account just now, not even associated with my work email
I'll try to get an azure subscription
![image](https://user-images.githubusercontent.com/689271/181570390-06aa7c3d-18a3-4dda-8afb-65442865c57e.png)

gzigzigzeo replied:
I've added the notice, anyway.

marcoandredinis commented on access/ms-teams/validate.go:
```diff
@@ -0,0 +1,136 @@
+	teamApp, err := b.GetTeamsApp(context.Background())
+	if trace.IsNotFound(err) {
+		fmt.Printf("Application %v not found in the org app store. Please, ensure that you have the application uploaded and installed for your team.", c.MSAPI.TeamsAppID)
+		return nil
```
I think we should return an error here :thinking: 
Otherwise it's _validated_

gzigzigzeo replied:
We do not want an error output and backtrace in this case. validate.go represents the console command which sends a fake message, it returns only fatal errors (networking, etc). Otherwise, it nicely shows the result to a user.

marcoandredinis replied:
Got it :+1: 

I would expect the `validate` method to return an error and we would then print the message on the caller
It's fine this way

marcoandredinis commented on access/ms-teams/bot.go:
```diff
@@ -0,0 +1,237 @@
+// user was not found.
+func (b Bot) FetchUser(ctx context.Context, userIDOrEmail string) (*UserData, error) {
+	if b.teamsApp == nil {
+		return nil, trace.Errorf("Bot is not configured, run GetTeamsApp first")
```
Can we just call `GetTeamsApp` here?

gzigzigzeo replied:
We could, yes, but the idea was to make user explicitly call `GetTeamsApp` and quit if it fails. It makes no sense to proceed further otherwise.

marcoandredinis commented on access/ms-teams/bot.go:
```diff
@@ -0,0 +1,237 @@
+	}
+
+	b.mu.RLock()
+	d, ok := b.users[userIDOrEmail]
```
Do we need to restart the plugin if we want a newly created user to receive notifications?

marcoandredinis commented on access/ms-teams/bot.go:
```diff
@@ -0,0 +1,237 @@
+// FetchUser fetches app id for user, installs app for a user if missing, fetches chat id and saves
+// everything to cache. This method is used for priming the cache. Returns trace.NotFound if a
+// user was not found.
+func (b Bot) FetchUser(ctx context.Context, userIDOrEmail string) (*UserData, error) {
```
Can we rename this method to something that indicates there's side effect when calling this method?

Something like `InstallAppForUser` or something similar?

gzigzigzeo replied:
This method does the following: 1) Tries to fetch user 2) Installs an app if missing 3) Gets the chat for a user. I do not think that `InstallAppForUser` fits. Do you probably have any ideas?

marcoandredinis replied:
`EnsureAppForUser`
Out of ideas today, sorry :joy: 

My point is that `Fetch` does not seem to indicate that we are actually installing the app
Maybe we can extract this to smaller functions

gzigzigzeo replied:
There are a smaller functions already, this method composes them into the strategy :-) Let's just leave it as is until I fix all other issues :-)

marcoandredinis commented on access/ms-teams/configure.go:
```diff
@@ -0,0 +1,164 @@
+		return trace.Wrap(err)
+	}
+
+	for _, a := range a {
```
We have two variables named `a`
Can we please rename one of them, if possible with something more explicit

marcoandredinis commented on access/ms-teams/configure.go:
```diff
@@ -0,0 +1,164 @@
+	}
+	defer w.Close()
+
+	err = tpl.ExecuteTemplate(w, "template", payload)
+	if err != nil {
+		return trace.Wrap(err)
+	}
```
```suggestion
	if err := tpl.ExecuteTemplate(w, "template", payload); err != nil {
		return trace.Wrap(err)
	}
```

gzigzigzeo replied:
Hmm, I use the multiline style almost everywhere, could you please elaborate a bit, why to use single line here?

marcoandredinis replied:
Sorry, it's ok
Ignore the other comments regarding this type of change

I usually prefer using the single line, but no need to change to it

marcoandredinis commented on access/ms-teams/main.go:
```diff
@@ -0,0 +1,105 @@
+	startCmd := app.Command("start", "Starts Teleport MS Teams plugin")
+	startConfigPath := startCmd.Flag("config", "TOML config file path").
+		Short('c').
+		Default("/etc/teleport-ms-teams.toml").
```
We should extract this to a const, it's used in multiple places

marcoandredinis commented on access/ms-teams/msapi/client.go:
```diff
@@ -0,0 +1,143 @@
+			return trace.Wrap(err)
+		}
+
+		if resp.StatusCode > 499 {
```
```suggestion
		if resp.StatusCode > 499 || resp.StatusCode == http.StatusTooManyRequests {
```

gzigzigzeo replied:
Thanks!

marcoandredinis commented on access/ms-teams/msapi/token.go:
```diff
@@ -0,0 +1,132 @@
+			return t, trace.Wrap(err)
+		}
+
+		if resp.StatusCode != http.StatusOK {
```
This never evaluates to true, does it?
We are `continue`ing at line 112 if `resp.StatusCode != http.StatusOK`

gzigzigzeo replied:
Right. Changed the order.

marcoandredinis commented on access/ms-teams/msapi/graph_client.go:
```diff
@@ -0,0 +1,265 @@
+		SuccessCode: http.StatusCreated,
+	}
+
+	err := c.request(ctx, request)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
```
```suggestion
	return trace.Wrap(c.request(ctx, request))
```

marcoandredinis commented on access/ms-teams/msapi/bot_framework_client.go:
```diff
@@ -0,0 +1,124 @@
+	m := postMessagePayload{
+		Type: "message",
+		From: postMessagePayloadFrom{
+			ID:   "29:` + botID + `",
```
We should move the `29` to a const and add an explanation (the link to the blog/article you talked about should do)

gzigzigzeo replied:
Okay, just in case here's the article: https://hackandchat.com/teams-proactive-messaging/

marcoandredinis replied:
I think we introduced a bug here
This will compile to
```
29:` + botID + `
```

Instead of `29:abcd`

gzigzigzeo replied:
Ouch, yes, surprisingly this works with a bug. My guess is that the API does not care.

gzigzigzeo replied:
Fixed.

marcoandredinis commented on access/ms-teams/msapi/bot_framework_client.go:
```diff
@@ -0,0 +1,124 @@
+		}},
+	}
+
+	body, err := json.MarshalIndent(&m, "", "    ")
```
Nothing against indenting the json, but is it required?

gzigzigzeo replied:
It facilitates the debugging

marcoandredinis commented on access/ms-teams/plugindata.go:
```diff
@@ -0,0 +1,61 @@
+		data.TeamsData = append(data.TeamsData, TeamsMessage{ID: channelID, Timestamp: timestamp})
+	}
+	if str := dataMap["messages"]; str != "" {
+		for _, encodedMsg := range strings.Split(str, ",") {
+			parts := strings.Split(encodedMsg, "/")
```
Can we assume that ID, Timestamp and RecipientID won't have neither a `,` or a `/`?

gzigzigzeo replied:
Yes. ID and RecipientID are both UUIDs, Timestamp is RFC822 if I am not mistaken. But you're right, this possibly needs changing. I've just used the same format we used in other plugins.

marcoandredinis commented on access/ms-teams/app.go:
```diff
@@ -0,0 +1,521 @@
+	apiClient, err := client.New(ctx, client.Config{
+		Addrs:       a.conf.Teleport.GetAddrs(),
+		Credentials: a.conf.Teleport.Credentials(),
+		DialOpts: []grpc.DialOption{
```
Let's add another DialOption here
```go
grpc.WithReturnConnectionError(),
```
https://github.com/gravitational/teleport-plugins/pull/601

marcoandredinis commented:
We should run the linter, I think it has some offenses

gzigzigzeo commented:
Okay, will do.

gzigzigzeo commented:
@Joerger @marcoandredinis Please, check the updated version of PluginData. Implemented generics and `Create`/`Update` instead of `Modify`. Refactored app.go a bit as well.

marcoandredinis commented on access/ms-teams/app.go:
```diff
@@ -0,0 +1,479 @@
+
+// onResolvedRequest is called when a request is resolved
+func (a *App) onResolvedRequest(ctx context.Context, req types.AccessRequest) error {
+	//	replyErr := a.postReviews(ctx, req.GetName(), req.GetReviews())
+
```
```suggestion

```

marcoandredinis commented on access/ms-teams/ms_teams_test.go:
```diff
@@ -0,0 +1,677 @@
+	require.NoError(t, err)
+	require.Equal(t, reviewer.ID, msg.RecipientID)
+
+	// max size of request was decreased here: https://github.com/gravitational/teleport/pull/13298
```
```suggestion
```

marcoandredinis commented on access/ms-teams/ms_teams_test.go:
```diff
@@ -0,0 +1,677 @@
+
+	require.Equal(t, msg.RecipientID, reply.RecipientID)
+	require.Equal(t, msg.ID, reply.ID)
+	require.Equal(t, gjson.Get(reply.Body, "attachments.0.content.body.4.facts.0.value").String(), "✅")
```
This paths look a bit cryptic
I think we should move them to consts with a clarifying name
What do you think?

gzigzigzeo replied:
XSLT of the new era :-) I agree, that would increase readability.

marcoandredinis commented:
PR looks very good, that's an awesome job you did here :clap: 
I still want to do a last pass, as we have a lot of changes
I'm still trying to get access to a Teams workplace to try out the plugin

Joerger commented on lib/plugindata/cas.go:
```diff
@@ -0,0 +1,166 @@
+	}
+}
+
+// TODO: Implement Upsert method
```
```suggestion
// Upsert tries to perform compare-and-swap update of a plugin data.
func (c *CompareAndSwap[T]) Upsert(ctx context.Context, resource string, upsertData T) (T, error) {
	emptyData := *new(T)

	backoff := backoff.NewDecorr(c.backoffBase, c.backoffMax, clockwork.NewRealClock())
	for {
		existingData, err := c.getPluginData(ctx, resource)
		if trace.IsNotFound(err) {
			existingData := emptyData
		} else if err != nil {
			return emptyData, trace.Wrap(err)
		}

		err = c.updatePluginData(ctx, resource, upsertData, existingData)
		if err == nil {
			return upsertData, nil
		} else if !trace.IsCompareFailed(err) {
			return emptyData, trace.Wrap(err)
		} else if err := backoff.Do(ctx); err != nil {
			return emptyData, trace.Wrap(err)
		}
	}
}```

gzigzigzeo replied:
Not exactly :-) We need callback and backoff because it is an extension to the update method. Let's follow YAGNI here: I update comment with "when it is required". As we are not using it right now, I do not see a good reason to add it.

Joerger commented on access/ms-teams/app.go:
```diff
@@ -0,0 +1,477 @@
+	pluginData, err := a.pd.Update(ctx, id, func(existing PluginData) (PluginData, error) {
+		teamsData := existing.TeamsData
+		if len(teamsData) == 0 {
+			return PluginData{}, trace.CompareFailed("existing teamsData is empty")
```
nit: clarifying comment
```suggestion
			// wait for the plugin data to be updated with TeamsData
			return PluginData{}, trace.CompareFailed("existing teamsData is empty")
```

Joerger commented on access/ms-teams/app.go:
```diff
@@ -0,0 +1,477 @@
+
+		count := len(reviews)
+		oldCount := existing.ReviewsCount
+		if oldCount >= count {
+			return PluginData{}, trace.CompareFailed("reviews are sent already")
+		}
```
If the reviews are already sent, won't the backoff attempts to CompareAndSwap also fail? Should we just return an error other than CompareFailed to abort the update?
```suggestion
		if oldCount >= count 
			return PluginData{}, trace.AlreadyExists("reviews are sent already")
		}
```

gzigzigzeo replied:
@Joerger Thank you for your thoroughness!

Joerger reviewed (commented):
Great use of generics!

gzigzigzeo commented on access/ms-teams/app.go:
```diff
@@ -0,0 +1,477 @@
+
+	// Let's try to create PluginData. This equals to locking AccessRequest to this
+	// instance of a plugin.
+	_, err := a.pd.Create(ctx, id, func() (PluginData, error) {
+		return PluginData{AccessRequestData: data}, nil
+	})
```
I agree, done.

gzigzigzeo commented on lib/plugindata/cas.go:
```diff
@@ -0,0 +1,166 @@
+	}
+}
+
+// Create tries to perform compare-and-swap update of a plugin data assuming that it does not exist
+//
+// fn callback function receives current plugin data value and returns modified value and
+// error.
+//
+// Please note that fn might be called several times due to CAS backoff, hence, you must be careful
+// with things like I/O ops and channels.
+func (c *CompareAndSwap[T]) Create(
+	ctx context.Context,
+	resource string,
+	cb func() (T, error),
+) (T, error) {
+	emptyData := *new(T)
+
+	existingData, err := c.getPluginData(ctx, resource)
+	if err != nil && !trace.IsNotFound(err) {
+		return emptyData, trace.Wrap(err)
+	}
+
+	if existingData != nil {
+		return emptyData, trace.AlreadyExists("plugin data already exists")
+	}
+
+	modifiedData, err := cb()
+	if err != nil {
+		return emptyData, trace.Wrap(err, "callback failed to update PluginData")
+	}
+
+	err = c.updatePluginData(ctx, resource, modifiedData, emptyData)
+	if err == nil {
+		return modifiedData, nil
+	}
+
+	return emptyData, trace.Wrap(err)
+}
```
Done

gzigzigzeo commented on lib/plugindata/cas.go:
```diff
@@ -0,0 +1,166 @@
+			return emptyData, trace.Wrap(err)
+		}
+
+		if oldData == nil {
+			return emptyData, trace.NotFound("Plugin data does not exist")
+		}
```
Done

gzigzigzeo commented on lib/plugindata/cas.go:
```diff
@@ -0,0 +1,166 @@
+
+// Update tries to perform compare-and-swap update of a plugin data assuming that it exist
+//
+// fn callback function receives current plugin data value and returns modified value and
+// error. If fn has failed to update data due to a diff, it must return CompareFailed error.
+func (c *CompareAndSwap[T]) Update(
+	ctx context.Context,
+	resource string,
+	cb func(T) (T, error),
+) (T, error) {
```
Done

gzigzigzeo commented on access/ms-teams/app.go:
```diff
@@ -0,0 +1,477 @@
+	log := logger.Get(ctx)
+
+	pluginData, err := a.pd.Update(ctx, reqID, func(existing PluginData) (PluginData, error) {
+		if len(existing.TeamsData) == 0 {
+			return PluginData{}, nil
+		}
+
+		// If resolution field is not empty then we already resolved the incident before. In this case we just quit.
+		if existing.AccessRequestData.ResolutionTag != pd.Unresolved {
+			return PluginData{}, nil
+		}
```
Sure, we should, fixed.

gzigzigzeo commented on lib/plugindata/cas.go:
```diff
@@ -0,0 +1,166 @@
+		expectData := *oldData
+
+		newData, err := cb(cbData)
+		if err != nil && !trace.IsCompareFailed(err) {
+			return emptyData, trace.CompareFailed("Callback failed to update PluginData")
+		}
```
It would not overwrite existing data because `updatePluginData` would not allow it. But you are correct, we should backoff immediately. Fixed.

gzigzigzeo commented:
@Joerger Thank you for your thouroughness!

Joerger commented on lib/plugindata/cas.go:
```diff
@@ -0,0 +1,166 @@
+		}
+		if !trace.IsCompareFailed(err) {
+			return emptyData, trace.Wrap(err)
+		}
```
Don't we still need the backoff here as well?
```suggestion
		}
		if err := backoff.Do(ctx); err != nil {
			return emptyData, trace.Wrap(err)
		}
```

Joerger reviewed (approved):
LGTM, assuming Marco's manual testing doesn't uncover any issues

marcoandredinis commented:
Just tested this in a real (it was actually a sandboxed) account and it works as expected :+1: 
![image](https://user-images.githubusercontent.com/689271/182176399-f1d07f90-6165-408e-be9b-1f2c67d48e7c.png)

I just have a little suggestion: increasing the timeout of the WatcherJob
The current value of 5s was not enough to start process the request, fetch user info and send the message
My suggestion is to set it to at least 15 seconds as default
Ideally this would be part of the configuration

A workaround for this is to turn on the `preload` in configuration

hugoShaka reviewed (approved):
LGTM, I would like to address a couple of thing but want to avoid the same giant PR than for the operator: I'll merge this PR with the typo changes and will send smaller/more reviewable PRs later.

hugoShaka commented on lib/plugindata/access_request.go:
```diff
@@ -0,0 +1,61 @@
+	ResolutionReason string
+}
+
+// DecodePluginData deserializes a string map to PluginData struct.
```
Fixed

hugoShaka commented on lib/plugindata/access_request.go:
```diff
@@ -0,0 +1,61 @@
+	return
+}
+
+// DecodePluginData deserializes a string map to PluginData struct.
```
Fixed

hugoShaka commented on access/ms-teams/msapi/client.go:
```diff
@@ -0,0 +1,143 @@
+type request struct {
+	// Method HTTP method
+	Method string
+	// Path path to a resource
```
Fixed

hugoShaka commented on access/ms-teams/_tpl/teleport-ms-teams-role.yaml:
```diff
@@ -0,0 +1,16 @@
+kind: user
```
Should specify the role first as `tctl create -f <filename>` will fail with: 
```
ERROR: role teleport-ms-teams is not found
```

hugoShaka commented on access/ms-teams/msapi/bot_framework_client.go:
```diff
@@ -0,0 +1,127 @@
+		Type: "message",
+		From: postMessagePayloadFrom{
+			ID:   botDesignator + botID,
+			Name: "TeleBot",
```
This is weird to hardcode this here: how does this behave with the app installed as an other name ?
